netket/nn/symmetric_linear.py

```diff
@@ -101,62 +101,446 @@ def full_bias(self, bias):
+    kernel_init: Optional[Callable[[PRNGKeyT, Shape, DType], Array]] = None
+    """Initializer for the Dense layer matrix."""
+    bias_init: Optional[Callable[[PRNGKeyT, Shape, DType], Array]] = None
+    """Initializer for the bias."""
 
-    The G-convolution generalizes the convolution to non-commuting groups:
+    def setup(self):
+        self.n_cells = np.product(np.asarray(self.shape))
+        self.n_symm = len(np.asarray(self.space_group)) // self.n_cells
+        self.sites_per_cell = np.asarray(self.space_group).shape[1] // self.n_cells
 
-    .. math ::
+        self.stdev = 1.0 / np.sqrt(self.n_cells * self.sites_per_cell)
 
-        C^i_g = \sum_h {\bf W}_{g^{-1} h} \cdot {\bf f}_h
+        self.mapping = (
+            np.asarray(self.space_group)[:, : self.sites_per_cell]
+            .reshape(self.n_cells, self.n_symm, self.sites_per_cell)
+            .transpose(1, 2, 0)
+            .reshape(self.n_symm, self.sites_per_cell, *self.shape)
+        )
 
-    Group elements that differ by the same symmetry operation (i.e. :math:`g = xh`
-    and :math:`g' = xh'`) are connected by the same filter.
+    def make_kernel(self, kernel):
+        kernel = kernel[..., self.mapping]
+
+        return kernel
+
+    @compact
+    def __call__(self, x: Array) -> Array:
+        """Applies the equivariant transform to the inputs along the last two
+        dimensions (-2: features, -1: group elements)
+        """
+
+        dtype = jnp.promote_types(x.dtype, self.dtype)
+        x = jnp.asarray(x, dtype)
+
+        x = (
+            x.reshape(-1, self.n_cells, self.sites_per_cell)
+            .transpose(0, 2, 1)
+            .reshape(-1, self.sites_per_cell, *self.shape)
+        )
+
+        if self.kernel_init:
+            kernel = self.param(
+                "kernel",
+                self.kernel_init,
+                (self.features, self.n_cells * self.sites_per_cell),
+                self.dtype,
+            )
+        else:
+            kernel = self.param(
+                "kernel",
+                normal(self.stdev),
+                (self.features, self.n_cells * self.sites_per_cell),
+                self.dtype,
+            )
+
+        kernel = jnp.asarray(kernel, dtype)
+
+        if self.mask:
+            kernel = kernel * jnp.expand_dims(self.mask, 0)
+
+        kernel = self.make_kernel(kernel)
+
+        x = jnp.fft.fftn(x, s=self.shape).reshape(*x.shape[:2], self.n_cells)
+
+        kernel = jnp.fft.fftn(kernel, s=self.shape).reshape(
+            *kernel.shape[:3], self.n_cells
+        )
+
+        x = lax.dot_general(
+            x, kernel, (((1,), (2,)), ((2,), (3,))), precision=self.precision
+        )
+        x = x.transpose(1, 2, 3, 0)
+        x = x.reshape(*x.shape[:3], *self.shape)
+
+        x = jnp.fft.ifftn(x, s=self.shape).reshape(*x.shape[:3], self.n_cells)
+        x = x.transpose(0, 1, 3, 2).reshape(*x.shape[:2], -1)
```

Comment:
Same here. Also, the FFT-based conv generates a lot of shared code between these two classes, consider separating them into a function?

Reply:
Note JAX doesn't (yet) offer an FFT-based conv, so we need to write it here

Reply:
There is also https://github.com/google/jax/pull/6343 which hopefully will eventually be merged.
I'd prefer having a single`DenseSymm` layer with `mode=direct/fft/conv/whatever` and the logic be in the function.

Reply:
> There is also [google/jax#6343](https://github.com/google/jax/pull/6343) which hopefully will eventually be merged.

Unlikely to work for us – the scipy version artificially removes the periodic BC option and doesn't work for complex numbers, these limitations will probably be inherited by the JAX version. Also, the contraction rules that follow are more elaborate than a simple broadcast

> I'd prefer having a single`DenseSymm` layer with `mode=direct/fft/conv/whatever` and the logic be in the function.

Could try, but the different implementations require different information about the group, so we'd be forced to pass `PermutationGroup` objects which slows down jitting **a lot**

Reply:
hmm...